hs_collection_resource/models.py

```diff
@@ -0,0 +1,140 @@
+        # no file types are supported
+        return ()
+
+    @classmethod
+    def can_have_multiple_files(cls):
+        # resource can't have any files
+        return False
+
+    @property
+    def metadata(self):
+        md = CollectionMetaData()
+        return self._get_metadata(md)
+
+processor_for(CollectionResource)(resource_processor)
+
+class Collection(AbstractMetaDataElement):
+    term = 'Collection'
+    resources = models.ManyToManyField(BaseResource, null=True, blank=True)
+
+    def get_resource_id_list(self):
+        return [res.short_id for res in self.resources.all()]
+
+    @classmethod
+    @transaction.atomic
+    def create(cls, **kwargs):
+        if 'resource_id_list' in kwargs:
+            metadata_obj = kwargs['content_object']
+            new_meta_instance = Collection.objects.create(content_object=metadata_obj)
+            for res_id in kwargs['resource_id_list']:
+                res = BaseResource.objects.filter(short_id__iexact=res_id)
+                if res.exists():
+                    new_meta_instance.resources.add(res[0])
+                else:
+                    raise ObjectDoesNotExist("No resource was found with the provided id: %s" % res_id)
+            return new_meta_instance
+        else:
+            raise ObjectDoesNotExist("No resource_id_list parameter was found in the **kwargs list")
+
+    @classmethod
+    @transaction.atomic
+    def update(cls, element_id, **kwargs):
+        meta_instance = Collection.objects.get(id=element_id)
+
+        if 'resource_id_list' in kwargs:
+            meta_instance.resources.clear()
+            for res_id in kwargs['resource_id_list']:
+                qs = BaseResource.objects.filter(short_id__iexact=res_id)
+                if qs.exists():
+                    meta_instance.resources.add(qs[0])
+                else:
+                    raise ObjectDoesNotExist("No resource was found with the provided id: %s" % res_id)
+            meta_instance.save()
+            if meta_instance.resources.all().count() == 0:
+                meta_instance.delete()
+        else:
+            raise ObjectDoesNotExist("No resource_id_list parameter was found in the **kwargs list")
+
+    @classmethod
+    def remove(cls, element_id):
+        raise ValidationError("Collection element can't be deleted.")
+
+
+class CollectionMetaData(CoreMetaData):
+
+    collection = GenericRelation(Collection)
+
+    @classmethod
+    def get_supported_element_names(cls):
+        elements = super(CollectionMetaData, cls).get_supported_element_names()
+        elements.append('Collection')
+        return elements
+
+    def has_all_required_elements(self):
+        if self.get_required_missing_elements():
+            return False
+        else:
+            # check if no member resource exists
+            if self.collection.first().resources.all().count() == 0:
+                return False
+            # check if all member resources are either public or private
```

Comment:
This comment probably is not right. Should it be ' check if all member resources are either public or discoverable'?